src/backend/distributed/planner/modify_planner.c

```diff
@@ -641,3 +644,315 @@ ExtractPartitionValue(Query *query, Var *partitionColumn)
+	}
+
+	/* make sure partition column is used only once in the query */
+	varClauseList = pull_var_clause_default(quals);
+	foreach(varClauseCell, varClauseList)
+	{
+		Var *column = (Var *) lfirst(varClauseCell);
+		if (equal(column, partitionColumn))
+		{
+			partitionColumnReferenceCount++;
+		}
+	}
+
+	if (partitionColumnReferenceCount != 1)
+	{
+		return false;
+	}
+
+	shardCount = ShardIntervalCount(distributedTableId);
+
+	if (shardCount == 0)
+	{
+		return false;
+	}
+
+	return true;
+}
+
+
+/* MultiRouterPlanCreate creates a physical plan for given router plannable query. */
+MultiPlan *
+MultiRouterPlanCreate(Query *query)
+{
+	Task *task = NULL;
+	Job *job = NULL;
+	MultiPlan *multiPlan = NULL;
+
+	task = SingleShardSelectTask(query);
+
+	job = SingleShardSelectJob(query,task);
+
+	multiPlan = CitusMakeNode(MultiPlan);
+	multiPlan->workerJob = job;
+	multiPlan->masterQuery = NULL;
+	multiPlan->masterTableName = NULL;
+
+	return multiPlan;
+}
+
+
+/* SingleShardSelectTask builds a Task to represent a single shard select query */
+static Task *
+SingleShardSelectTask(Query *query)
+{
+	Task *task = NULL;
+	ShardInterval *shardInterval = SingleShardSelectShardInterval(query);
+	StringInfo queryString = makeStringInfo();
+	uint64 shardId = INVALID_SHARD_ID;
+	bool upsertQuery = false;
+	CmdType commandType = query->commandType;
+	FromExpr *joinTree = NULL;
+
+	Assert(shardInterval != NULL);
+	Assert(commandType == CMD_SELECT);
+
+	shardId = shardInterval->shardId;
+
+	/*
+	 * Convert the qualifiers to an explicitly and'd clause, which is needed
+	 * before we deparse the query.
+	 */
+	joinTree = query->jointree;
+	if ((joinTree != NULL) && (joinTree->quals != NULL))
+	{
+		Node *whereClause = joinTree->quals;
+		if (IsA(whereClause, List))
+		{
+			joinTree->quals = (Node *) make_ands_explicit((List *) whereClause);
+		}
+	}
+
+	deparse_shard_query(query, shardInterval->relationId, shardId, queryString);
+	ereport(DEBUG4, (errmsg("distributed statement: %s", queryString->data)));
+
+	task = CitusMakeNode(Task);
+	task->jobId = INVALID_JOB_ID;
+	task->taskId = INVALID_TASK_ID;
+	task->taskType = ROUTER_TASK;
+	task->queryString = queryString->data;
+	task->anchorShardId = shardId;
+	task->dependedTaskList = NIL;
+	task->upsertQuery = upsertQuery;
+
+	return task;
+}
+
+
+/*
+ * SingleShardSelectShardInterval determines the single shard targeted by a
+ * provided distributed modification command. If no matching shards exist, or
+ * if the modification targets more than one one shard, this function raises
+ * an error.
+ */
+static ShardInterval *
+SingleShardSelectShardInterval(Query *query)
+{
+	List *restrictClauseList = NIL;
+	List *prunedShardList = NIL;
+	Index tableId = 1;
+
+	Oid distributedTableId = ExtractFirstDistributedTableId(query);
+	List *shardIntervalList = NIL;
+
+	/* error out if no shards exist for the table */
+	shardIntervalList = LoadShardIntervalList(distributedTableId);
+	if (shardIntervalList == NIL)
```

Comment:
Haven't we already check `shardCount` in `MultiRouterPlannableQuery ` ? I think this should be an `Assert`.